src/server/container_utils.h

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022, Roman Gershman.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+#pragma once
+
+extern "C" {
+#include "redis/intset.h"
+#include "redis/object.h"
+#include "redis/redis_aux.h"
+#include "redis/util.h"
+}
+
+#include <type_traits>
+
+#include "base/flags.h"
+#include "core/compact_object.h"
+#include "core/string_set.h"
+#include "server/common.h"
+
+ABSL_DECLARE_FLAG(bool, use_set2);
+
+namespace dfly {
+
+namespace container_utils {
+
+// StorageKind specifies how the container stores values.
+enum class StorageKind {
+  // Container stores all values as strings.
+  STRING,
+  // Container stores all values as integral values (int).
+  INTEGRAL,
+  // Container might store strings and integers.
+  MIXED
+};
+
+// StorageType is the respective C++ type of the container internals
+// to accommodate any of its values in the most optimal way.
+// STRING = std::string
+// INTEGRAL = int64_t
+// MIXED = std::variant<std::string, int64_t>
+template <StorageKind K>
+using StorageType =
+    std::conditional_t<K == StorageKind::MIXED, std::variant<std::string, int64_t>,
+                       std::conditional_t<K == StorageKind::STRING, std::string, int64_t>>;
+
+// IsContainer returns true if the iterator points to a container type.
+bool IsContainer(PrimeIterator it) {
```

Comment:
non template functions require inline in header files

Reply:
here and everywhere

Reply:
This is just a prototype so far to check the general approach. I'll look at details later, once its set